llm/turbine_llm/layers/kv_cache.py

```diff
@@ -0,0 +1,248 @@
+# Copyright 2024 Advanced Micro Devices, Inc
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+"""Light-weight encapsulations for various forms of attention KV-caches.
+
+These are not complete abstractions: they are primarily focused on making
+tightly coupled transformer blocks a bit less "stringy" with loose tensors
+and dims floating around everywhere. 
+"""
+
+import math
+
+import torch
+
+from ..utils.debugging import trace_tensor
+
+__all__ = [
+    "PagedKVCache",
+]
+
+
+class PagedKVCache:
+    """Implementation of a KV cache on top of a 'page table'.
+
+    The page table slab is physically represented as a 2D tensor:
+        [page_count, flattened_dims]
+
+    Each "page" can be thought of as a 6D view onto:
+
+    * transformer block
+    * cache partition (K or V cache)
+    * block sequence stride (number of sequence positions per block)
+    * attention heads
+    * attention dimensionality
+
+    Note that the internal page structure matches the organization of the
+    model, allowing contiguous individual local reads and writes at a sub-block
+    granularity if indexing deeply into the structure.
+    """
+
+    def __init__(
+        self,
+        *,
+        transformer_block_count: int,
+        attn_head_count: int,
+        attn_head_dim: int,
+        cache_partition_count: int = 2,
+        block_seq_stride: int = 16,
+    ):
+        self.transformer_block_count = transformer_block_count
+        self.attn_head_count = attn_head_count
+        self.attn_head_dim = attn_head_dim
+        self.cache_partition_count = cache_partition_count
+        self.block_seq_stride = block_seq_stride
+
+        # Some derived values based on attributes.
+        self.sub_page_dims = [
+            self.transformer_block_count,
+            self.cache_partition_count,
+            self.block_seq_stride,
+            self.attn_head_count,
+            self.attn_head_dim,
+        ]
+        self.page_slab_flat_dim = math.prod(self.sub_page_dims)
+
+    def unflatten_page_table(self, state: list[torch.Tensor]) -> torch.Tensor:
+        """Unflattens the 2D page table to a 6D tensor."""
+        assert len(state) == 1, f"Expected 1-element state. Got: {len(state)}"
+        page_slab = state[0]
+        return page_slab.reshape(
```

Comment:
Probably NIT, but any thoughts on using view to guarantee no redundant copy?

Reply:
I tend to use reshape unless if setting up for a mutation, which is the case that you need to be assured it is not a copy.

View imposes other costs for the guarantees it makes (around contiguity), which makes it a more tricky op in my view (no pun intended).

No free lunches...

Reply:
Thanks for the explanation, sg!